drivers/bluetooth/hci/rpmsg.c

```diff
@@ -0,0 +1,426 @@
+	data += sizeof(hdr);
+	remaining -= sizeof(hdr);
+
+	if (remaining != hdr.len) {
+		BT_ERR("Event payload length is not correct");
+		return NULL;
+	}
+	remaining = hdr.len;
+	BT_DBG("len %u", hdr.len);
+
+	buf = bt_buf_get_evt(hdr.evt, false, K_NO_WAIT);
+	if (!buf) {
+		BT_ERR("No available event buffers!");
+		return buf;
+	}
+
+	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	bt_buf_set_prio(buf, bt_hci_evt_is_prio(hdr.evt));
+
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static struct net_buf *bt_rpmsg_acl_recv(u8_t *data, size_t remaining)
+{
+	struct bt_hci_acl_hdr hdr;
+	struct net_buf *buf;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enought data for ACL header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ACL_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		remaining -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+		bt_buf_set_prio(buf, false);
+	} else {
+		BT_ERR("No available ACL buffers!");
+	}
+
+	if (remaining != hdr.len) {
+		BT_ERR("ACL payload length is not correct");
+		return NULL;
+	}
+	remaining = sys_le16_to_cpu(hdr.len);
+
+	BT_DBG("len %u", remaining);
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static void bt_rpmsg_rx(u8_t *data, size_t len)
+{
+	u8_t pkt_indicator;
+	struct net_buf *buf = NULL;
+	size_t remaining = len;
+
+	BT_HEXDUMP_DBG(data, len, "RPMsg data:");
+
+	pkt_indicator = *data++;
+	remaining -= sizeof(pkt_indicator);
+
+	switch (pkt_indicator) {
+	case RPMSG_EVT:
+		buf = bt_rpmsg_evt_recv(data, remaining);
+		break;
+
+	case RPMSG_ACL:
+		buf = bt_rpmsg_acl_recv(data, remaining);
+		break;
+
+	default:
+		BT_ERR("Unknown HCI type %u", pkt_indicator);
+		return;
+	}
+
+	if (buf) {
+		BT_DBG("Calling bt_recv(%p)", buf);
+		if (bt_buf_get_prio(buf)) {
+			bt_recv_prio(buf);
+		} else {
+			bt_recv(buf);
+		}
+
+		BT_HEXDUMP_DBG(buf->data, buf->len, "RX buf payload:");
+	}
+}
+
+static int bt_rpmsg_send(struct net_buf *buf)
+{
+	u8_t pkt_indicator;
+
+	BT_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_OUT:
+		pkt_indicator = RPMSG_ACL;
+		break;
+	case BT_BUF_CMD:
+		pkt_indicator = RPMSG_CMD;
+		break;
+	default:
+		BT_ERR("Unknown type %u", bt_buf_get_type(buf));
+		goto done;
+	}
+	net_buf_push_u8(buf, pkt_indicator);
+
+	BT_HEXDUMP_DBG(buf->data, buf->len, "Final HCI buffer:");
+	rpmsg_send(&ep, buf->data, buf->len);
+
+done:
+	net_buf_unref(buf);
+	return 0;
+}
+
+static int bt_rpmsg_open(void)
+{
+	int err;
+
+	BT_DBG("");
+
+	err = bt_rpmsg_init_internal();
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+static const struct bt_hci_driver drv = {
+	.name		= "RPMsg",
+	.open		= bt_rpmsg_open,
+	.send		= bt_rpmsg_send,
+};
+
+static int bt_rpmsg_init(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	int err;
+
+	err = bt_hci_driver_register(&drv);
+	if (err) {
+		return err;
+	}
+
+	return 0;
```

Comment:
This could all be replaced with a single line `return bt_hci_driver_register(&drv)`, unless you plan to put more stuff into this function in the future.

Reply:
simplified